homeassistant/components/stream/core.py

```diff
@@ -42,17 +64,69 @@ class Segment:
     stream_id: int = attr.ib(default=0)
-    parts: list[Part] = attr.ib(factory=list)
+    # Parts are stored in a dict indexed by http range start for easy
+    # lookup when requested by byterange
+    # As of Python 3.7, insertion order is preserved, and we insert
+    # in sequential order, so the Parts are ordered
+    parts_by_http_range: dict[int, Part] = attr.ib(factory=dict)
     start_time: datetime.datetime = attr.ib(factory=datetime.datetime.utcnow)
+    # Store text of this segment's hls playlist for reuse
+    hls_playlist: str = attr.ib(default=None)
+    hls_playlist_parts: str = attr.ib(default=None)
+    # Number of playlist parts rendered so far
+    hls_num_parts_rendered: int = attr.ib(default=0)
+    # Set to true when all the parts are rendered
+    hls_playlist_complete: bool = attr.ib(default=False)
 
     @property
     def complete(self) -> bool:
         """Return whether the Segment is complete."""
         return self.duration > 0
 
+    @property
+    def data_size_with_init(self) -> int:
+        """Return the size of all part data + init in bytes."""
+        return len(self.init) + self.data_size_without_init
+
+    @property
+    def data_size_without_init(self) -> int:
+        """Return the size of all part data without init in bytes."""
+        # We can use the last part to quickly calculate the total data size.
+        if not self.parts_by_http_range:
+            return 0
+        last_part = next(reversed(self.parts_by_http_range.values()))
+        return last_part.http_range_start + len(last_part.data)
+
     def get_bytes_without_init(self) -> bytes:
         """Return reconstructed data for all parts as bytes, without init."""
-        return b"".join([part.data for part in self.parts])
+        return b"".join([part.data for part in self.parts_by_http_range.values()])
+
+    def get_part_bytes(self, start_loc: int) -> bytes:
+        """Return part that begins at start_loc by looking up index in the part map.
+
+        Just a helper method for the remaining_data method below.
+        """
+        part = self.parts_by_http_range.get(start_loc)
+        return b"" if part is None else part.data
+
+    def get_aggregating_bytes(
+        self, start_loc: int, end_loc: int | float
+    ) -> Generator[bytes, None, None]:
+        """Yield available remaining data until segment is complete or end_loc is reached.
+
+        Begin at start_loc. End at end_loc (exclusive).
+        Used to help serve a range request on a segment.
+        """
+        pos = start_loc
+        # Since we use this from a non worker thread, we need to check complete before
+        # checking for new data. Use | instead of "or" to avoid short circuit evaluation.
+        while (not self.complete) | bool(bytes_to_write := self.get_part_bytes(pos)):
+            pos += len(bytes_to_write)
+            if pos >= end_loc:
+                assert isinstance(end_loc, int)
```

Comment:
Hmm, I'm not sure what we should do. Assertions should be disabled when the optimize flag is passed to the interpreter, but I guess we are not passing that parameter for HA. Anyway our use is just an artifact of typing. @KapJI what is the best practice here?

Reply:
I think asserts are fine if they're checking some predicate which must be true.

In this particular case it's not clear to me why `end_loc` only can be an `int` if function is typed to accept `float`. Maybe it shouldn't accept `float`?

Alternatively, the value can be explicitly converted to `int` with `int(end_loc)`.

Reply:
The only float we use is infinity, because float has infinity and int doesn't. Therefore if `pos` >= `end_loc`, `end_loc` must not be infinity and thus must be an `int`.

Reply:
I think `sys.maxsize` can be used as a reasonable replacement for `float("inf")`.

Reply:
I suppose so. It's a huge number. But that's just the platform's pointer size, and it's not really a bound on ints (sys.maxsize+1 is a fine number). There's no reason this constant should be tied to platform pointer size really. We might as well just define any very large number and not tie it to the system.
There actually is a large number (2^^53-1) suggested in RFC8643 that I was originally using for this. We will most likely never see that number. But using it is a hack nonetheless - infinity is cleaner at least mathematically.
Anyway, I'll make the change to a defined int if we determine that assert is really to be avoided.

Reply:
I think i was mistaken about assert, apologies. I found plenty of uses and probably misread something.